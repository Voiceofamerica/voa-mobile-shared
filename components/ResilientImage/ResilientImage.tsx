
import * as React from 'react'
import { Subscription } from 'rxjs/Subscription'

import { waitUntilOnline } from 'helpers/resilience'

import { resilientImage, imageContent } from './ResilientImage.scss'

export interface Props {
  src: string
  className?: string
  alwaysShow?: boolean
  style?: React.CSSProperties
  defaultSrc?: string
  onLoadDone?: () => void
}

export type ImageStatus = 'loading' | 'loaded' | 'error'

export interface State {
  shouldRender?: boolean
  imageStaus: ImageStatus
}

const makeAttemptUrl = (src: string) => `${src}#${Date.now()}`
const noop = () => null
const IMAGE_FETCH_RETRY_RATE = 2000

class ReilientImage extends React.Component<Props, State> {
  state: State = {
    imageStaus: 'loading',
  }

  componentWillMount () {
    const shouldRender = navigator.onLine || this.props.alwaysShow
    this.setState({
      shouldRender,
    })

    if (shouldRender) {
      this.tryFetchImage()
    }
  }

  tryFetchImage = () => {
    if (!navigator.onLine) {
      return waitUntilOnline().then(() => this.tryFetchImage())
    }

    const { onLoadDone = noop } = this.props

    const img = new Image()

    img.addEventListener('error', () => {
      this.setImageStatus('error')
      setTimeout(this.tryFetchImage, IMAGE_FETCH_RETRY_RATE)
    })

    img.addEventListener('load', () => {
      this.setImageStatus('loaded')
      onLoadDone()
    })

    img.src = this.props.src
  }

  setImageStatus = (imageStaus: ImageStatus) =>
    this.setState({
      imageStaus,
    })

  renderOffline = () => (
    <div style={{ width: 0, height: 0 }} />
  )

  renderContent = () => {
    const { imageStaus } = this.state
    const { src, defaultSrc = require('./imagedefault.gif') } = this.props

    const srcToUse = imageStaus === 'loaded'
                   ? src
                   : defaultSrc

    return (<img src={srcToUse} className={imageContent} />)
  }

  render () {
    const { shouldRender = navigator.onLine } = this.state
    if (!shouldRender) {
      return this.renderOffline()
    } else {
      const {
        className = '',
        style,
      } = this.props

      return (
        <div className={`${resilientImage} ${className}`} style={style}>
          { this.renderContent() }
        </div>
      )
    }
  }
}

export default ReilientImage
