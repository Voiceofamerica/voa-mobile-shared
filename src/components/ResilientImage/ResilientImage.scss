
@import '../../commonStyles/index.scss';

.resilientImage {
  position: relative;
  background: black;

  .imageContent {
    object-fit: cover;
    width: 100%;
    height: 100%;

    &.containImage {
      object-fit: contain;
      height: auto;
      max-height: 100%;
    }
  }

  .childrenArea {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    display: flex;
    justify-content: center;
    align-items: center;
  }

  .spinner {
    color: $voa-blue;
    height: 80%;
    max-height: 100px;
    margin-top: -1%;
  }
}
