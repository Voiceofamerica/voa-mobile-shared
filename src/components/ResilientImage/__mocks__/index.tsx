
import * as React from 'react'

export default jest.fn(({ children, ...props}) => (<div {...props}>{children}</div>))
