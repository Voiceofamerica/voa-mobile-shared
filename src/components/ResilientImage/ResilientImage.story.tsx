
import * as React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import ResilientImage from './ResilientImage'

const imageUrl = 'https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w650_r1.png'

storiesOf('ResilientImage', module)
  .add('display', () => (
    <ResilientImage src={imageUrl} />
  ))
  .add('display without image', () => (
    <ResilientImage src='asdf' />
  ))
