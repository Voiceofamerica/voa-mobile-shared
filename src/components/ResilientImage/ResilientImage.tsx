
import * as React from 'react'

import { waitUntilOnline } from '../../helpers/resilience'

import Spinner from '../Spinner'

import { resilientImage, imageContent, childrenArea, containImage, spinner } from './ResilientImage.scss'

export interface Props {
  src: string | null
  showSpinner?: boolean
  className?: string
  alwaysShow?: boolean
  style?: React.CSSProperties
  defaultSrc?: string
  onLoadDone?: () => void
  onClick?: () => void
  contain?: boolean
}

export type ImageStatus = 'loading' | 'loaded' | 'error'

export interface State {
  shouldRender?: boolean
  imgUrl?: string
  imageStaus: ImageStatus
}

const noop = () => null
const IMAGE_FETCH_RETRY_RATE = 2000

class ReilientImage extends React.Component<Props, State> {
  state: State = {
    imageStaus: 'loading',
  }

  mounted: boolean = false

  readonly abortController = new AbortController()

  componentWillMount () {
    this.mounted = true
    const shouldRender = navigator.onLine || this.props.alwaysShow
    this.setState({
      shouldRender,
    })

    if (shouldRender) {
      this.tryFetchImage()
    }
  }

  componentWillUnmount () {
    this.mounted = false
    this.abortController.abort()
  }

  componentWillReceiveProps (nextProps: Props) {
    if (nextProps.src !== this.props.src) {
      this.setImageStatus('loading')

      const shouldRender = navigator.onLine || this.props.alwaysShow
      this.setState({
        shouldRender,
      })

      if (shouldRender) {
        this.tryFetchImage()
          .catch(err => {
            console.warn(`Something went wrong loading ${nextProps.src}`, err)
          })
      }
    }
  }

  tryFetchImage = async () => {
    const { src } = this.props

    if (!src || !this.mounted) {
      return
    }

    if (!navigator.onLine) {
      waitUntilOnline().then(() => this.mounted && this.tryFetchImage())
      return
    }

    const { onLoadDone = noop } = this.props

    try {
      const blobUrl = await fetch(src, {
          method: 'GET',
          signal: this.abortController.signal,
        })
        .then(res => res.blob())
        .then(blob => URL.createObjectURL(blob))

      this.setImageStatus('loaded', blobUrl)
      onLoadDone()

    } catch (err) {
      this.setImageStatus('error')
      setTimeout(this.tryFetchImage, IMAGE_FETCH_RETRY_RATE)
    }
  }

  setImageStatus = (imageStaus: ImageStatus, imgUrl?: string) => {
    if (!this.mounted) {
      return
    }

    this.setState({
      imageStaus,
      imgUrl,
    })
  }

  renderOffline = () => (
    <div style={{ width: 0, height: 0 }} />
  )

  renderContent = () => {
    const { imageStaus, imgUrl } = this.state
    const { defaultSrc = require('./imagedefault.gif'), contain } = this.props

    const isLoaded = imageStaus === 'loaded'
    const srcToUse = isLoaded && imgUrl
                   ? imgUrl
                   : defaultSrc

    const className = contain ? `${imageContent} ${containImage}` : imageContent

    return (<img src={srcToUse} className={className} />)
  }

  renderSpinner = () => {
    const { showSpinner } = this.props
    const { imageStaus } = this.state

    const isLoaded = imageStaus === 'loaded'

    if (isLoaded || !showSpinner) {
      return null
    }

    return (
      <div className={childrenArea}>
        <Spinner className={spinner} />
      </div>
    )
  }

  renderChildren = () => {
    const { children } = this.props

    return (
      <div className={childrenArea}>
        {children}
      </div>
    )
  }

  render () {
    const { shouldRender = navigator.onLine || this.props.alwaysShow } = this.state
    if (!shouldRender) {
      return this.renderOffline()
    } else {
      const {
        className = '',
        style,
        onClick = () => null,
      } = this.props

      return (
        <div className={`${resilientImage} ${className}`} style={style} onClick={ev => onClick()}>
          { this.renderContent() }
          { this.renderSpinner() }
          { this.props.children ? this.renderChildren() : null }
        </div>
      )
    }
  }
}

export default ReilientImage
