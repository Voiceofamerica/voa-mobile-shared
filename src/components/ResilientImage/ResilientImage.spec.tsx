
import * as React from 'react'
import { create } from 'react-test-renderer'
import { shallow } from 'enzyme'

import ResilientImage from './ResilientImage'

jest.mock('../Spinner')

describe(`<${ResilientImage.name} />`, () => {
  afterAll(() => {
    Object.defineProperty(window.navigator, 'onLine', { value: true })
  })
  describe('shapshots', () => {
    it('should render the default image if alwaysShow is true', () => {
      const element = create((
        <ResilientImage src='https://some.url/something.png' alwaysShow />
      ))

      expect(element).toMatchSnapshot()
    })
    it('should render a zero-size div if offline', () => {
      Object.defineProperty(window.navigator, 'onLine', { value: false })
      const element = create((
        <ResilientImage src='https://some.url/something.png' />
      ))

      expect(element).toMatchSnapshot()
    })
  })
})
