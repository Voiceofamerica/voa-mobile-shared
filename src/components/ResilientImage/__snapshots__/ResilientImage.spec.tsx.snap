// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResilientImage /> shapshots should render a zero-size div if offline 1`] = `
<div
  style={
    Object {
      "height": 0,
      "width": 0,
    }
  }
/>
`;

exports[`<ResilientImage /> shapshots should render the default image if alwaysShow is true 1`] = `
<div
  className="resilientImage "
  onClick={[Function]}
  style={undefined}
>
  <img
    className="imageContent"
    src="Some File"
  />
</div>
`;
